### PR TITLE
Rolling upgrade code

### PR DESCRIPTION
**Pull request by bodepd**

### CODE REVIEW

**Comment by anshprat** [comment]
checking on this now..

**Comment by anshprat** [comment]
@bodepd with consul now having the control flags, we need to have logging of these data changes. Not necessarily in scope of this PR though. Ideas?

**Comment by bodepd** [comment]
@anshprat I think that is way too much data to store in consul. I think it would make more sense to store the changes on disk.

**Comment by anshprat** [comment]
@bodepd oh yes, definitely on a log file only on disk and not in consul.

**Comment by sorenh** [comment]
It seems problematic that while an upgrade is going on, Puppet is disabled. For large clusters, this could take a long time. If the upgrade somehow fails, it could take a very, very long time. During this time, the pending nodes will have no mechanism for self-healing.

Thoughts?

File: README.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+    {
+      "group_mappings":{
+        "foo":"bar"
+      }
+    }
+
+NOTE: I am not sure that I like this feature, it is really complicated
+and effects overall usability by making it less then obvious what rules
+a given host will map to. I am considering elimating it, it should only
```
Comment (sorenh) [inline README.md]: "eliminating"

**Comment by bodepd** [comment]
@sorenh this solution isn't intended to be perfect, just good enough so that we can start playing with it. I think completely disabling Puppet is reasonable for the first iteration (b/c it's the easiest way to ensure that everything works like we expect). The solution that I really want is to have meaningful version numbers that we can set as keys (but this won't be possible until package snapshotting is done).

We could also allow healing runs to happen even if this key is set, but it might lead to cases where machines upgrade accidentally (is this true, this seems to be what people are saying)?

File: README.md [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+* version - version that should be set for current\_version
+* instructions - instrcutions for how to coordinate when hosts can upgrade
+
+*Version* is used to set current version, if it does not match the current version,
+it also runs global\_puppet\_disable to put everything in the default pending state.
+
+*Instructions* - json that can be passed in from the command line or read from the config
+file */etc/jio-upgrade.json*
+
```
Comment (vpramo) [inline README.md]: @bodepd should this not be jiocloud-upgrade.json

**Comment by anshprat** [comment]
long due - did a quick test in vagrant today - the older functions work as earlier, so merging now.

**Comment by bodepd** [comment]
@anshprat hahahahaha

**Pull request merged**